tool/tctl/common/device/resource.go

```diff
@@ -0,0 +1,154 @@
+// Copyright 2023 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package device
+
+import (
+	"time"
+
+	"github.com/gravitational/trace"
+	"google.golang.org/protobuf/types/known/timestamppb"
+
+	devicepb "github.com/gravitational/teleport/api/gen/proto/go/teleport/devicetrust/v1"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/devicetrust"
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+// Resource is a wrapper around devicepb.Device that implements types.Resource.
+type Resource struct {
+	// ResourceHeader is embedded to implement types.Resource
+	types.ResourceHeader
+	// Spec is the device specification
+	Spec ResourceSpec `json:"spec"`
+}
+
+// ResourceSpec is the device resource specification.
+// This spec is intended to closely mirror `devicepb.Device` but swaps some data around
+// to get a UX that matches with our other resource types.
+type ResourceSpec struct {
+	// OsType is represented as a string here for user-friendly manipulation.
+	OsType      string                      `json:"os_type"`
+	AssetTag    string                      `json:"asset_tag"`
+	CreateTime  time.Time                   `json:"create_time,omitempty"`
+	UpdateTime  time.Time                   `json:"update_time,omitempty"`
+	EnrollToken *devicepb.DeviceEnrollToken `json:"enroll_token,omitempty"`
+	// OsType is represented as a string here for user-friendly manipulation.
+	EnrollStatus  string                          `json:"enroll_status"`
+	Credential    *devicepb.DeviceCredential      `json:"credential,omitempty"`
+	CollectedData []*devicepb.DeviceCollectedData `json:"collected_data,omitempty"`
+}
+
+// checkAndSetDefaults sanity checks Resource fields to catch simple errors, and
+// sets default values for all fields with defaults.
+func (r *Resource) checkAndSetDefaults() error {
+	if err := r.Metadata.CheckAndSetDefaults(); err != nil {
+		return trace.Wrap(err)
+	}
+	if r.Kind == "" {
+		r.Kind = types.KindDevice
+		// Sanity check.
+	} else if r.Kind != types.KindDevice {
+		return trace.BadParameter("unexpected resource kind %q, must be %q", r.Kind, types.KindDevice)
+	}
+
+	if _, err := devicetrust.ResourceOSTypeFromString(r.Spec.OsType); err != nil {
```

Comment:
Is there a point in running the enum conversions here, just do it again on `resourceToProto`? Unless I misread, one method always follows the other, as both are effectively a part of Unmarshaling.

Reply:
Nothing aside from convention. I suppose. I've pretty closely followed convention to how our other resources work out of API familiary. If someone calls this function in the future, they'd expect it to make sure the resource is fully valid.